---
title: "Tutorials for longitudinal analysis of gene or protein expression data"
author: "Ozan Aygun"
date: "June 13, 2017"
output:
  html_document:
    depth: 4
    highlight: tango
    number_sections: yes
    theme: cerulean
    toc: yes
  pdf_document:
    toc: yes
---


```{r setup, include=FALSE, fig.align='center',fig.width=5,fig.height=4}
knitr::opts_chunk$set(echo =TRUE, include = TRUE, message = FALSE,warning = FALSE,fig.align='center',fig.width=5,fig.height=4)
```

___

# Introduction and Summary

# Longitudinal analysis using timecourse package

Here our interest is to analyze data that is obtained in the form of gene expression raios, to identify genes that are differentially expressed in a time-course, replicated experiment design that involves one or more biological conditions.

**timecourse** package (deposited in bioconductor) provides a nice platform for handling these type of data. The workhorse functions are **mb.long()** and **mb.MANOVA()**. These functions calculate MB or T-squared statistics using **multivariate emprical Bayes approaches.**

- The input is a matrix of gene (or protein) expression ratios that are properly preprocessed. log2 expression ratios are suitable.

The statistical advantage of this approach (over the traditional F-statistic) is that:

- it incorporates replicate variances
- it accounts for the correlation among the gene expression measurements across the time-points
- it also performs **moderation** that is, borrowing information across the the genes present in the analysis to shrink the standard error estimates towards a common value. This is especially useful in the analyses that involve small number of independent samples and helps to make robust inferences about the coefficients (i.e: increases the statistical power).

## Longitudinal one-sample problem and its analysis

The data set we are going to use here is the **fruitfly** data set that comes with the package:

```{r}
library(timecourse)
data("fruitfly")
head(fruitfly)
```

This data set is a matrix of log2 gene expression values for 2000 probesets from Drosophila. The experimental design consists of RNA isolated from fly embryos at 1h to 12h post laying the eggs.

- microarray measuments
- 3 biological replicates (A, B, C)
- Each replicate set has 12 time points (01,02,..,12)

Let's look at the distribution of individual experiments:

```{r, fig.height=6}
color.map <- colorRampPalette(c("blue", "red", "green"))(36)
plot(density(fruitfly[,1]), type = "l", col = color.map[1], ylim = c(0,0.25),
     main = "Distribution of expression ratios", xlab = "log2 expression")
for(i in 2:ncol(fruitfly)) lines(density(fruitfly[,i]), col = color.map[i])
```

We realize that the data is not scaled or centered, but has reasonably normal distribution for all experiments. Their medians and means are also quite similar. This could have important consequences if the data was not preprocessed properly.

Note that this is not a true longitudinal design, because there are not repeated measurements from the same individual, rather there are seperate embryos grown for each time point. However, since we can assume (or actually suspect) that the embryos that are used for the time course have very similar characteristics, we can model these results as longitudinal data. This could provide better inference relative to simply not accounting for any correlation between the time points and simply performing a factorial design.

## Question: which genes significantly change over time?

We are trying to answer this simple question in this one-sample scenario. 

### Define and assign the time point and group information for each experiment

Here what we refer as experiment is simply one data column. The function **mb.long()** assumes that the data (expression) matrix is arranged in the following order:

- **Biological Conditions (Groups) -> Replicates -> Time points**

For the current problem, since we have just one group (e.g: Wild-type Drosophila embryos), the function will assume that the columns are arranged in the order of:

- **Replicates -> Time points**

Let's see what this means:

- If 3 replicates we have are labelled as: (A, B, C)
- If each time point is labelled as: (01,02,..,12)

Then, the column order that is required for the **mb.long()** function is:

**A01,A02,...,A12,B01,B02,...,B12,C01,C02,...,C12**

This is quite intuitive. 

In this example, we are lucky and don't need to perform any column shuffling. The default order in this example data set is in this desired order:

```{r}
colnames(fruitfly)
```


### Use mb.long() to calculate MB and T-squared statictics

Note that: 

- **times**:the number of time points and 
- **reps**:  A numeric vector or matrix corresponding to the sample sizes for all genes across different biological conditions

are always required as input arguments.

- **method**: defaults to "1D": for the one-sample case where genes of interest are those which change over time. 

```{r,fig.width=8, fig.height=6}
size <- rep(3,2000) # we have 3 replicates each have 2000 data points (i.e: genes)
# times will be set to 12 (the number of time points for each replicate)

out1 <- mb.long(fruitfly, times = 12, reps = size)
plotProfile(out1)
```

This is an example of how timeseries package can be used to determine the genes of interest that change over time for a given group. Note that timeseries package just provides statistics and ranks the genes. It does NOT however, perform any hypothesis testing, therefore does NOT provide p-values. While this is not a big problem, often researchers are interested in getting a threshold to shortlist the genes that are significantly changed over the time. Therefore, we continue learning additional approaches that handle longitudinal data analysis for gene expression sets.

# Longitudinal analysis using LIMMA package and duplicateCorrelation() function

In this example we will have 2 problems to work with:

1. One - group longitudinal analysis to find the genes that are changing over the time
2. Two - group longitudinal analysis to find the genes that are changing over the time and between the groups

For this tutorial, I will use a dummy data set that contains log2 expression values we obtained from 12 experiments. 

The experimental design is:

- There are WT and Mutant cells
- There is one treatment
- There are 3 time points for each treatment, for each cell line
- We have 2 biological replicate (independent) measurements for each time point, , for each cell line

Therefore, the true independent sample size for each gene, and each time point (assuming there are no missing values) is 2. This implies that we will need to perform moderation for the covariance matrix using the information across the genes in order to make more robust inferences.

In this set up, the protein expression measurements within WT and Mut groups are correlated and we need to account for this correlation.

Within LIMMA framework, this could be performed by using the Multi-level-model paradigm and introducing **random effects** for the units of repeated measurements.

In this example, we will include random effects for WT or Mutant cells as they are the **units of correlation.** 

We will start with reading and properly structuring the data, then fitting one-group, and two-group comparison models and interpreting the results.


## Reading the data and annotation matrix

In order to perform the 


