---
title: "Tutorials for longitudinal analysis of gene or protein expression data"
author: "Ozan Aygun"
date: "June 13, 2017"
output:
  html_document:
    depth: 4
    highlight: tango
    number_sections: yes
    theme: cerulean
    toc: yes
  pdf_document:
    toc: yes
---


```{r setup, include=FALSE, fig.align='center',fig.width=5,fig.height=4}
knitr::opts_chunk$set(echo =TRUE, include = TRUE, message = FALSE,warning = FALSE,fig.align='center',fig.width=5,fig.height=4)
```

___

# Introduction and Summary

# Longitudinal analysis using timecourse package

Here our interest is to analyze data that is obtained in the form of gene expression raios, to identify genes that are differentially expressed in a time-course, replicated experiment design that involves one or more biological conditions.

**timecourse** package (deposited in bioconductor) provides a nice platform for handling these type of data. The workhorse functions are **mb.long()** and **mb.MANOVA()**. These functions calculate MB or T-squared statistics using **multivariate emprical Bayes approaches.**

- The input is a matrix of gene (or protein) expression ratios that are properly preprocessed. log2 expression ratios are suitable.

The statistical advantage of this approach (over the traditional F-statistic) is that:

- it incorporates replicate variances
- it accounts for the correlation among the gene expression measurements across the time-points
- it also performs **moderation** that is, borrowing information across the the genes present in the analysis to shrink the standard error estimates towards a common value. This is especially useful in the analyses that involve small number of independent samples and helps to make robust inferences about the coefficients (i.e: increases the statistical power).

## Longitudinal one-sample problem and its analysis

The data set we are going to use here is the **fruitfly** data set that comes with the package:

```{r}
library(timecourse)
data("fruitfly")
head(fruitfly)
```

This data set is a matrix of log2 gene expression values for 2000 probesets from Drosophila. The experimental design consists of RNA isolated from fly embryos at 1h to 12h post laying the eggs.

- microarray measuments
- 3 biological replicates (A, B, C)
- Each replicate set has 12 time points (01,02,..,12)

Let's look at the distribution of individual experiments:

```{r, fig.height=6}
color.map <- colorRampPalette(c("blue", "red", "green"))(36)
plot(density(fruitfly[,1]), type = "l", col = color.map[1], ylim = c(0,0.25),
     main = "Distribution of expression ratios", xlab = "log2 expression")
for(i in 2:ncol(fruitfly)) lines(density(fruitfly[,i]), col = color.map[i])
```

We realize that the data is not scaled or centered, but has reasonably normal distribution for all experiments. Their medians and means are also quite similar. This could have important consequences if the data was not preprocessed properly.

Note that this is not a true longitudinal design, because there are not repeated measurements from the same individual, rather there are seperate embryos grown for each time point. However, since we can assume (or actually suspect) that the embryos that are used for the time course have very similar characteristics, we can model these results as longitudinal data. This could provide better inference relative to simply not accounting for any correlation between the time points and simply performing a factorial design.

## Question: which genes significantly change over time?

We are trying to answer this simple question in this one-sample scenario. 

### Define and assign the time point and group information for each experiment

Here what we refer as experiment is simply one data column. The function **mb.long()** assumes that the data (expression) matrix is arranged in the following order:

- **Biological Conditions (Groups) -> Replicates -> Time points**

For the current problem, since we have just one group (e.g: Wild-type Drosophila embryos), the function will assume that the columns are arranged in the order of:

- **Replicates -> Time points**

Let's see what this means:

- If 3 replicates we have are labelled as: (A, B, C)
- If each time point is labelled as: (01,02,..,12)

Then, the column order that is required for the **mb.long()** function is:

**A01,A02,...,A12,B01,B02,...,B12,C01,C02,...,C12**

This is quite intuitive. 

In this example, we are lucky and don't need to perform any column shuffling. The default order in this example data set is in this desired order:

```{r}
colnames(fruitfly)
```


### Use mb.long() to calculate MB and T-squared statictics

Note that: 

- **times**:the number of time points and 
- **reps**:  A numeric vector or matrix corresponding to the sample sizes for all genes across different biological conditions

are always required as input arguments.

- **method**: defaults to "1D": for the one-sample case where genes of interest are those which change over time. 

```{r,fig.width=8, fig.height=6}
size <- rep(3,2000) # we have 3 replicates each have 2000 data points (i.e: genes)
# times will be set to 12 (the number of time points for each replicate)

out1 <- mb.long(fruitfly, times = 12, reps = size)
plotProfile(out1)
```

This is an example of how timeseries package can be used to determine the genes of interest that change over time for a given group. Note that timeseries package just provides statistics and ranks the genes. It does NOT however, perform any hypothesis testing, therefore does NOT provide p-values. While this is not a big problem, often researchers are interested in getting a threshold to shortlist the genes that are significantly changed over the time. Therefore, we continue learning additional approaches that handle longitudinal data analysis for gene expression sets.

# Longitudinal analysis using LIMMA package and duplicateCorrelation() function

In this example we will have 2 problems to work with:

1. One - group longitudinal analysis to find the genes that are changing over the time
2. Two - group longitudinal analysis to find the genes that are changing over the time and between the groups

For this tutorial, I will use an example data set that contains log2 expression values we obtained from 12 experiments. The expression measurements are in the form of log2 ratios, which are relative to a common reference denominator. Individual experiments are centered at zero using median normalization.

The experimental design is:

- There are WT and Mutant cells
- There is one treatment
- There are 3 time points for each treatment (0h, 3h, 6h), for each cell line
- We have 2 biological replicate (independent) measurements for each time point,for each cell line

Therefore, the true independent sample size for each gene, and each time point (assuming there are no missing values) is 2. This implies that we will need to perform moderation for the covariance matrix using the information across the genes in order to make more robust inferences.

In this set up, the protein expression measurements within WT and Mut groups are correlated and we need to account for this correlation.

Within LIMMA framework, this could be performed by using the Multi-level-model paradigm and introducing **random effects** for the units of repeated measurements.

In this example, we will include random effects for WT or Mutant cells (i.e: the **subject** level) as they are the **units of correlation.** 

We will start with reading and properly structuring the data, then fitting one-group, and two-group comparison models and interpreting the results.


## Reading the data and annotation matrix

In order to perform the longitudinal analysis, we need to export a data matrix, along with an annotation matrix that matches the columns of this data matrix to the attibutes necessary to correctly fit the model.

Read the data matrix:

```{r}
data.mat <- readRDS("log_expression.rds")
head(data.mat)
```

Annotation table can be prepared in an R environment or externally using excel and then imported into the R session like I did here.

Read and look at the structure of the table:

```{r}
annotations <- read.csv("annotations.csv", stringsAsFactors = TRUE)
annotations
```

Note that the annotation matrix is composed of the following columns:

- **column.name:** explicitly matches the data columns to their attributes
- **Subject:** the subject ID for the repeated measure unit. Note that we have 2 independent replicates for this experiment. Therefore, we have 2 independent subjects for each of the Wt and Mut groups. The correlation exists between the time-series measurements within these subjects, therefore we will include random effects for the subjects.
- **Repeated.measure:** in this example, the repeated measure is the time points, 0,3 and 6h.
- **Group:** signifies which experimental group the experiment belongs to.

## One-group longitudinal analysis

In this analysis we will just investigate the question: **which genes change significantly over the time in Wt cells?**

We need to get the relevant part of the annotation matrix and experiment columns from the data matrix to pursue the analysis.

```{r}


annotations.wt <- annotations[annotations$Group == "Wt",]
annotations.wt

data.mat.wt <- data.mat[,which(colnames(data.mat) == annotations.wt$Column.Name)]
```

We therefore have 2 Subjects (independent biological replicates), 3 time-points (repeated measures) of gene expression measurements.

### Combine repeated measure and group levels

At this stage, it is convenient to combine the repeated measure levels and group levels:

```{r}
measure.levels <- paste(annotations.wt[,"Group"], annotations.wt[,'Repeated.measure'], sep='.')
measure.levels
```

### Prepare the design matrix

The design matrix for lmFit is prepared using the measure.levels we generated. Note the way we set up the levels for the measure levels, such that the reference level (in this case the zero time point) is the first and comparison measurement levels follow:

```{r}
f <- factor(measure.levels, levels = c("Wt.0","Wt.3","Wt.6"))
design <- model.matrix(~0 + f)
colnames(design) <- levels(f)
design
```

Note that for each independent experiment represents a parameter is for a given repeated measure level. The design matrix is most similar to the one-way multiple group comparisons we performed in limma, but in this case we are dealing with repeated measures data.

Recall that the order of the rows of the design matrix should match the columns of the data:

```{r}
names(data.mat.wt);annotations.wt$Column.Name
row.names(design) <- colnames(data.mat.wt)
design
```

The design matrix is now more intuitive.

### Using duplicateCorrelation() to estimate the correlation between the subjects

The purpose of this step is to account for the correlation between the gene expression measurements made on the same **Subject**. We will use the **duplicateCorrelation()** function:

This function minimally requires the following arguments:

- **object**: the data matrix that contains the repeated gene expression measurements for the Wt group
- **design**: the design matrix we prepared above, where rows corresponding to data object columns (experiments) and columns correspond to comparisons to be estimated. 
- **block**: vector or factor specifying a blocking variable. In other words, this is the **Subject** variable from the annotation matrix that contains the vector of subjects. In LIMMA, this is referred to as **blocking variable**. The blocking variable contains repeated observations that are correlated. Typically the blocks are biological replicates and the repeated observations are technical replicates that are correlated within a block. In our case, the repeated observations are the **time points.** 

**How are the correlations estimated?**

The correlation is estimated by **fitting a mixed linear model by REML individually for each gene**. 

```{r}
library(limma)
corfit <- duplicateCorrelation(data.mat.wt,design = design, block = annotations.wt$Subject)
```


### Using consensus correlation to fit the model by lmFit()

The duplicateCorrelation() function returns a **consensus correlation**, which is **a robust average of the individual correlations**, which can be used as input for functions lmFit or gls.series.

```{r}
corfit$consensus.correlation
```

Then, this inter-subject correlation is input into the linear model fit:

```{r}
fit <- lmFit(data.mat.wt, design = design, block = annotations.wt$Subject, correlation = corfit$consensus.correlation)
```

We now have the model fit accouting for the correlation between the repeated measures. The next step would be preparing the relevant contrasts to answer our question.

### Prepare the contrasts of interest: compare to baseline measurement level

In this case our contrasts are designed so that we are testing the significance of changes at each time point relative to the baseline measurement:

```{r}
contr.baseline <- makeContrasts(
        Wt_3vs0 = Wt.3 - Wt.0,
        Wt_6vs0 = Wt.6 - Wt.0,
        levels = design)
contr.baseline # This is the contrast matrix for baseline comparisons
```

### Prepare the contrasts of interest: compare each measure to the previous measure

This could be another set of comparisons of interest and we can easily prepare them:

```{r}
contr.previous <- makeContrasts(
        Wt_3vs0 = Wt.3 - Wt.0,
        Wt_6vs3 = Wt.6 - Wt.3,
        levels = design)
contr.previous # This is the contrast matrix for baseline comparisons
```

### Compute the contrasts

This is performed with contrasts.fit(): 

```{r}
fit.previous <- contrasts.fit(fit, contrasts = contr.previous)
fit.baseline <- contrasts.fit(fit, contrasts = contr.baseline)
```

### Perform emprical Bayes moderation

We perform the moderation by using the eBayes() function:

```{r}
fit.previous <- eBayes(fit.previous)
fit.baseline <- eBayes(fit.baseline)
```

### Get the results for individual contrasts

Fpr a given contrast,it is possible to extract coefficients (estimates) and p-values (inferences) using the topTable() function:

```{r}
results.previous <- topTable(fit.previous, coef = 2, number = nrow(data.mat.wt),adjust.method = "fdr", sort.by = "none")
results.baseline <- topTable(fit.baseline, coef = 2, number = nrow(data.mat.wt),adjust.method = "fdr", sort.by = "none")

library(scales)
plot(results.baseline$P.Value,results.previous$P.Value, col = alpha("navy",0.2), pch = 20)

```

Note that the inferences that are made relative to previous measure or to the base line could be very different for a given gene.

### Overall model inference (F-based p-value)

Often the question of interest is **whether a given gene display significant change over time**, regardless of when the change happened. This question relates to the inference made about the overall model and can be extracted from the **$F.p.value** slot of the fitted model object.

```{r}
plot(fit.baseline$F.p.value,fit.previous$F.p.value,col = alpha("navy",0.1), pch = 20)
```

Note that regardless of the contrasts we fitted, the overall model p-value is the same.

However, note that these F.p.values are not adjusted for multiple testing and we need to correct for that:

```{r}
adj.p.val <- p.adjust(fit.baseline$F.p.value, method = "fdr")
sum(adj.p.val < 0.05)
```

Note that in this case none of the genes display an overall significant change across the time of repeated measurements.

### Plot the N most significant genes that display change over time

We can derive different sets of time-series plots. The most generic ones would be ranking the genes based on their F.p.values and plotting a handful of most significant ones.

First of all, we need to get the data into long format and match the relevant annotations:

```{r}
library(reshape)
data.mat.wt$id <- row.names(data.mat.wt)
data.mat.wt.long <- melt(data.mat.wt, id = "id") # melt the data, keep id columns as the id
colnames (data.mat.wt.long) <- c('id', 'Column.Name', 'measurement')
# Merge the long data with the experiment annotations, match each column name with its attributes
data.mat.wt.long.mannt <- merge(data.mat.wt.long, annotations.wt, by = "Column.Name") 

```

We then extract the topN genes of interest:

```{r}
N = 10 # let's say we are intrested in the 10 most sifnificant genes

gene.fit <- cbind.data.frame(id = data.mat.wt$id, F.p.value = fit.baseline$F.p.value, adj.p.val)
gene.fit <- gene.fit[order(gene.fit$F.p.value),]
gene.fit.TopN <- gene.fit[1:10,]
```

Finally we get the data in long format for the topN most significant genes:

```{r}
data.m <- merge(data.mat.wt.long.mannt, gene.fit.TopN)
data.m$Subject <- factor(data.m$Subject)
```

then, we can plot the profiles:

```{r, fig.height= 12}
library(ggplot2)

ggplot (aes (x=Repeated.measure, y=measurement, group=Subject, color=Subject), data=data.m) + 
      geom_point() + geom_line() + facet_grid(id ~ .) + theme_bw() 
 
# Save the plot as pdf
 ggsave (paste ('limma-top', N, '.pdf', sep=''), width=4, height=1*N, pointsize=8)
```

This completes the one-group analysis on the longitudinal gene expression data.


## Two-group longitudinal analysis


Setting up the two-group analysis is very similar to the one-group one, and again depending on the question, contrasts need to be carefully defined. If the question of interest relates to the repeated.measure x group interaction, the relevant interaction term needs to be fitted in the form of contrasts, and p-value associated with that contrast needs to be used for the inference.

We will focus on the same data set, but this time we are interested in the following question: **which genes respond DIFFERENTLY over time in the mutant compared to wild-type?**

This is a question that relates to **difference of differences**, therefore can be reduced to interaction term.

Recall our annotation table:

```{r}
knitr::kable(annotations, align = "c", format = "pandoc")
```

We have 2 groups that have all the repeated measures. Each repeated measure is blocked by 2 independent biological replicates, or subjects.

Our approach will be very similar to a one-group analysis, except when we set our contrasts.

### Combine repeated measure and group labels:

```{r}
measure.levels.two.group <- paste(annotations$Group,annotations$Repeated.measure, sep = ".")
measure.levels.two.group
```

### Prepare the design matrix:

```{r}
f.two.group <- factor(measure.levels.two.group, levels = c( "Wt.0" , "Wt.0" , "Wt.3",  "Wt.3",  "Wt.6" , "Wt.6"  ,"Mut.0", "Mut.0" ,"Mut.3" ,"Mut.3", "Mut.6", "Mut.6"))
design.two.group <- model.matrix(~0 + f.two.group)
design.two.group
```
